webrender/src/device/gfx.rs

```diff
@@ -577,48 +517,47 @@ impl<B: hal::Backend> Image<B> {
         &mut self,
         device: &mut B::Device,
         cmd_pool: &mut hal::CommandPool<B, hal::queue::Graphics>,
+        staging_buffer_pool: &mut BufferPool<B>,
         rect: DeviceUintRect,
         layer_index: i32,
         image_data: &[u8],
-        offset_alignment: usize,
     ) -> hal::command::Submit<B, hal::Graphics, hal::command::OneShot, hal::command::Primary>
     {
-        //let (image_width, image_height, _, _) = self.kind.dimensions();
         let pos = rect.origin;
         let size = rect.size;
-        let offset = self.upload_buffer.update(device, image_data, offset_alignment);
+        staging_buffer_pool.add(device, image_data);
+        let buffer = staging_buffer_pool.buffer();
         let mut cmd_buffer = cmd_pool.acquire_command_buffer(false);
 
         let range = hal::image::SubresourceRange {
             aspects: hal::format::Aspects::COLOR,
             levels: 0 .. 1,
             layers: layer_index as _ .. (layer_index + 1) as _,
         };
-        if let Some(barrier) = self.core.transit(
-            hal::image::Access::TRANSFER_WRITE,
-            hal::image::Layout::TransferDstOptimal,
-            range.clone(),
-        ) {
+        let mut barriers = Vec::new();
+        barriers.extend(buffer.transit(hal::buffer::Access::TRANSFER_READ));
+        barriers.extend(
+            self.core.transit(
+                hal::image::Access::TRANSFER_WRITE,
+                hal::image::Layout::TransferDstOptimal,
+                range.clone(),
+            )
+        );
+        if !barriers.is_empty() {
             cmd_buffer.pipeline_barrier(
                 PipelineStage::COLOR_ATTACHMENT_OUTPUT .. PipelineStage::TRANSFER,
                 hal::memory::Dependencies::empty(),
-                &[barrier],
+                &barriers,
             );
         }
-
-        let buffer_width = if self.kind.extent().width == size.width {
-            self.upload_buffer.buffer.row_pitch() as u32 / self.format.bytes_per_pixel()
-        } else {
-            size.width
-        };
         cmd_buffer.copy_buffer_to_image(
-            &self.upload_buffer.buffer.buffer,
+            &buffer.buffer,
             &self.core.image,
             hal::image::Layout::TransferDstOptimal,
             &[
                 hal::command::BufferImageCopy {
-                    buffer_offset: self.upload_buffer.offset,
-                    buffer_width,
+                    buffer_offset: (staging_buffer_pool.offset - staging_buffer_pool.size) as _,
```

Comment:
hmm, did you mean `-size.width` here?

Reply:
The `buffer_pool` offset is already pointing to the new position where will add the next copy. And the size is stores the actual copy size. So `offset - size` will give us the offset we need for this copy.